api-report/container-runtime.api.md

```diff
@@ -496,7 +495,7 @@ export interface ISummarizer extends IEventProvider<ISummarizerEvents>, IFluidRo
     run(onBehalfOf: string, options?: Readonly<Partial<ISummarizerOptions>>): Promise<SummarizerStopReason>;
     // (undocumented)
     stop(reason: SummarizerStopReason): void;
-    summarizeOnDemand(options: IOnDemandSummarizeOptions): OnDemandSummarizeResult;
+    summarizeOnDemand(options: IOnDemandSummarizeOptions): Promise<OnDemandSummarizeResult>;
```

Comment:
From consumer POV, return result is just a monster! It's a promise of promise, with some conditional stuff sandwiched in between.
Can we simplify? As a consumer, I feel like I'd prefer one of the two models (depending on what exactly I want to accomplish):
1. An API that always accomplishes my request, even if it needs to want for current summary to finish and start on-demand summary.
2. An API that fails (exception) or maybe returns undefined if it fails due to concurrent summary, otherwise returns me result (just ISummarizeResults, no Promise<ISummarizeResults>, one can use SummarizeResultBuilder to construct it to wrap existing async activity).

Reply:
I like option 2, maybe just returning undefined if we're already summarizing.

Reply:
summarizeOnDemand does need to be async, though, if it's going to await RunningSummarizer.start. The alternative is to make the user call summarizer.start() manually in order to use it, which I'm not crazy about. Or include summarizer.start() in the work that requestSummarizer() does, which would be better. But I would ultimately like to have a single async method that spins up a summarizer, runs it, tears it down, and returns the result as Promise<ISummarizeResults>. That will work well with our partner's future plans.

Reply:
That's fine, but we can make API better. Currently it's essentially Promise<Promise<something>> (because result is just a collection of promioses), and it's a bit weird. That's the reason there Arin built building blocks to help here - SummarizeResultBuilder. You can synchronously create and return result but use continuations to correctly fail or complete those deferrals later.

But I think my bigger concern is around shape from different angle. I.e. as a user, it's weird to get Promise<void> telling me that some other summary is in process. I think it should either internally wait for it to finish and run on-demand summary, or just throw (and maybe there is a boolean argument passed into this function to differentiate caller preference between those two paths). I.e. Word would prefer failure, as in their server-driven summaries they will be in control to ensure there are no overlapping summaries. Probably same in tests, where test would want to disable heuristics-based summary and just drive it manually.

Reply:
Right. I'm making the change to remove the alreadyRunning promise. I still have summarizeOnDemand defined as an async function because of the async startup of the RunningSummarizer that it needs to wait for. I could try to add the RunningSummarizer startup promise to ISummarizeResults if you think that's worth it.